### PR TITLE
Add section on one-shot transform

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Address @trbromley 's [comment](https://github.com/PennyLaneAI/pennylane/pull/5266#issuecomment-1980957742).

**Description of the Change:**
Add section in `measurements.rst` introducing `dynamic_one_shot`. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: doc/introduction/measurements.rst [hunk 1]
```diff
@@ -300,6 +300,37 @@ The decorator syntax applies equally well:
...
         (...)
 
+The one-shot transform
+**********************
+
+Devices supporting mid-circuit measurements (defined using
+:func:`~.pennylane.measure`) and conditional operations (defined using
+:func:`~.pennylane.cond`) natively can estimate dynamic circuits by executing
+them one shot at the time. This is the default behaviour of a QNode that has a
+device supporting mid-circuit measurements, as well as any QNode with the
```
Comment (trbromley) [inline doc/introduction/measurements.rst:310]: This is only when in finite-shot mode right? And we're using deferred for analytic still?

File: doc/introduction/measurements.rst [hunk 2]
```diff
@@ -300,6 +300,37 @@ The decorator syntax applies equally well:
...
+.. code-block:: python
+
+    @qml.dynamic_one_shot
+    @qml.qnode(dev)
+    def my_quantum_function(x, y):
+        (...)
+
+.. warning::
+
+    The resulting function is not differentiable.
```
Comment (trbromley) [inline doc/introduction/measurements.rst:332]: Can we not use the parameter shift method?

**Review by trbromley — APPROVED** [review]
Nice, thanks for thinking about updating the docs - a crucial part of the user experience!

We should eventually think about pulling some of this content out into, e.g., a separate quickstart for dynamic circuits, as the measurements quickstart has started to feel a bit heavy.

**Review by mudit2812 — COMMENTED** [review]
Looks good! One note that should get addressed, otherwise I'm happy to approve.

File: doc/introduction/measurements.rst (hunk 2 above)
Comment (josh146) [reply doc/introduction/measurements.rst:332]: We can if the `dynamic_one_shot` transform is creating a one-shot tape that terminates in `expval`, but not `sample`

Comment (trbromley) [reply doc/introduction/measurements.rst]: Thanks! I think PS is also only broken when there is postselection in the MCM circuit, and maybe in some more advanced cases like having parametrized gates in while loops, but for most of the time parameter shift should be ok.

Perhaps those edge cases (postselection, etc.) should just be handled as a validation step in the parameter shift gradient transform, and then we could remove the warning here.

Comment (vincentmr) [reply doc/introduction/measurements.rst]: How about `Conditional` gates, wouldn't that break PS too? 

> Perhaps those edge cases (postselection, etc.) should just be handled as a validation step in the parameter shift gradient transform, and then we could remove the warning here.

Would you like that done in this PR, or some follow-up?

File: doc/introduction/measurements.rst [hunk 3]
```diff
@@ -300,6 +300,44 @@ The decorator syntax applies equally well:
...
+The one-shot transform
+**********************
+
+Devices supporting mid-circuit measurements (defined using
+:func:`~.pennylane.measure`) and conditional operations (defined using
+:func:`~.pennylane.cond`) natively can estimate dynamic circuits by executing
+them one shot at a time. This is the default behaviour of a QNode that has a
+device supporting mid-circuit measurements, as well as any QNode with the
+:func:`~.pennylane.dynamic_one_shot` quantum function transform.
+As the name suggests, this transform only works for QNode executing with finite shots
```
Comment (mudit2812) [inline doc/introduction/measurements.rst:312]:
```suggestion
them one shot at a time. This is the default behaviour of a `~.pennylane.QNode` that has a
device supporting mid-circuit measurements, as well as any `~.pennylane.QNode` with the
:func:`~.pennylane.dynamic_one_shot` quantum function transform.
As the name suggests, this transform only works for a `~.pennylane.QNode` executing with finite shots
```

**Review by mudit2812 — APPROVED** [review]

File: doc/introduction/measurements.rst (hunk 2 above)
Comment (trbromley) [reply doc/introduction/measurements.rst]: > How about Conditional gates, wouldn't that break PS too?

We can formalize this as a discovery item in Q2, but conditional gates generally don't break parameter-shift.

> Would you like that done in this PR, or some follow-up?

Yes, that can be a follow up, maybe once we have more certainty in Q2 about what does and doesn't work.

File: doc/introduction/measurements.rst [hunk 4]
```diff
@@ -300,6 +300,44 @@ The decorator syntax applies equally well:
...
+    @qml.dynamic_one_shot
+    @qml.qnode(dev)
+    def my_quantum_function(x, y):
+        (...)
+
+.. warning::
+
+    Dynamic circuits executed with shots should be differentiated with the finite-difference method.
+    If the ``defer_measurements`` transform is used in analytic mode, ``backprop`` is also a viable
+    option.
```
Comment (trbromley) [inline doc/introduction/measurements.rst:339]: Sorry @vincentmr, in retrospect would you mind adding "should be differentiated with the *parameter-shift or* finite-difference methods."

We can also link to https://docs.pennylane.ai/en/stable/introduction/interfaces.html#gradients

Comment (vincentmr) [reply doc/introduction/measurements.rst:339]: Should we instead say `finite-difference or backprop when post-selection is not None on some MCM`?

Comment (trbromley) [reply doc/introduction/measurements.rst:339]: Let's talk offline, I'm a little unsure what to do here :thinking: Probably lean towards removing this message and just having clear warnings in the gradient methods.

**Pull request merged**